### ISSUE TITLE
Not clear what system identity is

### ISSUE THREAD

**Issue by stepancheg** [bevyengine/bevy#10225]
Code example.

```
use bevy::app::App;
use bevy::app::Startup;

fn system1() {
    println!("system1");
}

fn main() {
    App::new().add_systems(Startup, (system1, system1)).run();
}
```

Prints "system1" twice. Which means, the function is instantiated as two systems, there's no identity, each function reference creates a new system.

But then instead I do:

```
App::new().add_systems(Startup, system1.before(system1)).run();
```

and now it panics at startup 

```
`"system1"` and `"SystemTypeSet(system1\")"` have both `in_set` and `before`-`after` relationships (these might be transitive). This combination is unsolvable as a system cannot run before or after a set it belongs to.
```

So, now bevy tracks uses function identity (pointer?) to detect cycles?

This is inconsistent. Either a function pointer is identity, and first example should print once (or emit some kind of error because the same system is being added twice), or second example should not panic and instantiate two systems.

Bevy 0.11.3.

**Comment by stepancheg** [comment]
OK, similar issue:

```
use bevy::app::App;
use bevy::app::Startup;
use bevy::prelude::IntoSystemConfigs;

fn system1() {
    println!("system1");
}

fn system2() {
    println!("system2");
}

fn system3() {
    println!("system3");
}

fn main() {
    App::new()
        .add_systems(Startup, (system3.after(system1), system3.after(system2)))
        .run();
}
```

Naively I think, that this means:
* system1 and system2 are executed concurrently
* then system3 is executed

But instead, system1 and system2 are not executed at all, but system3 is executed twice.

Would be nice if Bevy could at least detect this and report it.

**Comment by mockersf** [comment]
> ```
>         .add_systems(Startup, (system3.after(system1), system3.after(system2)))
> ```
> 
> But instead, system1 and system2 are not executed at all, but system3 is executed twice.

This is expected. `system3.after(system1)` means "add a constraint that `system3` must run once `system1` is done", and the `.after(system1)` does not schedule `system1`.

So the schedule looks at what it has to execute, there's no `system1` or `system2` left to execute, and it executes `system3` twice

**Comment by pnarimani** [comment]
I don't think anyone intentionaly would write `system3.after(system1)` and not add `system1`. It would be great if we add warnings to help with it. 
Same with adding the same system twice. I can't think of a use case where someone would want to add the same system twice.

**Comment by stepancheg** [comment]
> I don't think anyone intentionally would write `system3.after(system1)` and not add system1
> Same with adding the same system twice.

Well, I did.

Because it is not clear what that operation is doing, I assumed it adds both and adds dependency link between them.

> I can't think of a use case where someone would want to add the same system twice.

Right. So graph consistency checks would help, i.e. no system is added twice, and all references are added to the same schedule.

**Comment by hymm** [comment]
We used to have warnings when a label was used, but not configured. But this led to issues when there's a dependency on an optional plugin. i.e. a system in Plugin A needed to be after a system in Plugin B, but only if plugin B is added. Warnings would need to correctly account for this scenario somehow.

Adding a system twice has to be supported. We do it with `apply_buffers` and users will readd transform propagation systems if they need to repropagate.

**Comment by stepancheg** [comment]
@hymm 


> Adding a system twice has to be supported. We do it with apply_buffers and users will readd transform propagation systems if they need to repropagate.

My 2 cents comment would be a suggestion to make it explicit. E.g.

```
app.add_systems(
  Update, (
    apply_buffers,
    new_id(apply_buffers),
  ),
)
```

**Comment by bushrat011899** [comment]
> Code example.
> 
> ```rust
> use bevy::app::App;
> use bevy::app::Startup;
> 
> fn system1() {
>     println!("system1");
> }
> 
> fn main() {
>     App::new().add_systems(Startup, (system1, system1)).run();
> }
> ```
> 
> Prints "system1" twice. Which means, the function is instantiated as two systems, there's no identity, each function reference creates a new system.

This, as described, is desired behaviour. There are many instances where a user would want to run the same system multiple times. A good example of this is the `apply_deferred` system, which ensures all deferred operations are applied to the `World`. It's very common to see `apply_deferred` added many times throughout the schedules. If you could not add a single system multiple times to the schedule, this would be really cumbersome (new-type the system for every addition?)

> But then instead I do:
> 
> ```rust
> App::new().add_systems(Startup, system1.before(system1)).run();
> ```
> 
> and now it panics at startup
> 
> ```
> `"system1"` and `"SystemTypeSet(system1\")"` have both `in_set` and `before`-`after` relationships (these might be transitive). This combination is unsolvable as a system cannot run before or after a set it belongs to.
> ```
> 
> So, now bevy tracks uses function identity (pointer?) to detect cycles?

Again, this makes sense. `system1` is added multiple times, so there are multiple instances of `system1` in the schedule. Asking the scheduler to ensure `a before a` is nonsensical. Just as nonsensical as if you new-typed `system1` to be `system1_b` and added the constraints `system1 before system1_b && system1_b before system1`.

> This is inconsistent. Either a function pointer is identity, and first example should print once (or emit some kind of error because the same system is being added twice), or second example should not panic and instantiate two systems.

I think you're confusing what `before` and `after` mean in this context: they are purely for ordering, they do not add any systems themselves. `a.before(b)` returns a single system, `a` with the condition `a before b`. 

> OK, similar issue:
> 
> ```rust
> use bevy::app::App;
> use bevy::app::Startup;
> use bevy::prelude::IntoSystemConfigs;
> 
> fn system1() {
>     println!("system1");
> }
> 
> fn system2() {
>     println!("system2");
> }
> 
> fn system3() {
>     println!("system3");
> }
> 
> fn main() {
>     App::new()
>         .add_systems(Startup, (system3.after(system1), system3.after(system2)))
>         .run();
> }
> ```
> 
> Naively I think, that this means:
> 
>     system1 and system2 are executed concurrently
>     then system3 is executed
> 

What that actually means is "add `system3` with the condition `system3 after system1`; add `system3` with the condition `system3 after system2`". If you wanted that behaviour, you would instead:

```rust
app.add_systems(Startup, (
    system1,
    system2,
    system3.after(system1).after(system2),
));
```

Notice that the `after` method is an _additive_ constraint. To call it twice is to add two conditions, not replace them.

> But instead, system1 and system2 are not executed at all, but system3 is executed twice.

Again, that is the intended behaviour.

> Would be nice if Bevy could at least detect this and report it.

Now this is something that I think makes sense to act on. As @hymm mentioned, it is desireable to mention labels (systems) in constraints which may not exist (either temporarily or permanently). I do think adding an optional warning similar to the system ambiguity detection would be a nice change.

```rust
// Hypothetical API
app.edit_schedule(Update, |schedule| {
    schedule.set_build_settings(ScheduleBuildSettings {
        // Warn if a label is used without initialization
        label_unitialization_detection: LogLevel::Warn,
        ..default()
    });
});
```

However, I think this is far enough removed from your initial issue it might be worth adding as its own succinct feature request.

**Comment by alice-i-cecile** [comment]
While this was productive, I think this is largely unactionable in its current form: like @bushrat011899 said, I think that following up with other clear feature requests or documentation suggestions is a good idea.
